[SPARK-20132][Docs] Add documentation for column string functions

map222 opened the pull request:
## What changes were proposed in this pull request?
Add docstrings to column.py for the Column functions `rlike`, `like`, `startswith`, and `endswith`. Pass these docstrings through `_bin_op`

There may be a better place to put the docstrings. I put them immediately above the Column class.

## How was this patch tested?

I ran `make html` on my local computer to remake the documentation, and verified that the html pages were displaying the docstrings correctly. I tried running `dev-tests`, and the formatting tests passed. However, my mvn build didn't work I think due to issues on my computer.

These docstrings are my original work and free license. 

@davies has done the most recent work reorganizing `_bin_op`

map222 committed:
Added documentation for like, rlike, startswith, and endswith

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented:
Fix the title please?

HyukjinKwon commented on python/pyspark/sql/column.py:
```diff
@@ -124,6 +124,35 @@ def _(self, other):
+                   >>> df.filter(df.name.endswith('ice$')).collect()
+                   []
+               '''
+startswith_doc = ''' Return a Boolean :class:`Column` based on a string match.\n
```
Mind adding `_` as a prefix in this variable to indicate this is a private one?

map222 replied:
Ah, thank you! I couldn't figure out how to get the docs from not showing up as separate functions, but this solved it. I added `_` to the front of the docstrings, and moved them into the `Column` class, immediately before the `_bin_op` definitions.

holdenk commented on python/pyspark/sql/column.py:
```diff
@@ -250,11 +250,39 @@ def __iter__(self):
+               >>> df.filter( df.name.like('Al%') ).collect()
+               [Row(name=u'Alice', age=1)]
+                """
+    _startswith_doc = ''' Return a Boolean :class:`Column` based on a string match.\n
```
This documentation is different than the Scala docs, we generally try and keep them the same in other places so it might make sense to update both.

map222 replied:
Are [these the Scala docs](http://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.sql.Column@rlike(literal:String):org.apache.spark.sql.Column) you're referring to? It looks like the Scala Column docs are missing examples for a couple more functions that are documented in python (substr, isin). I could expand the ticket / PR to include Scala examples for all of these functions. I also noticed isNull and isNotNull are not document either.

holdenk reviewed (commented):
Thanks so much for working on this, improved Python documentation super important! I've got some initial comments :)

map222 committed:
Fixed formatting issues: indents, and block quotes

map222 commented on python/pyspark/sql/column.py:
```diff
@@ -250,11 +250,39 @@ def __iter__(self):
+
+               >>> df.filter( df.name.like('Al%') ).collect()
+               [Row(name=u'Alice', age=1)]
+                """
```
I made the indentation more consistent, and changed all the block quotes to `"""` to match the rest of the file in most recent commit.

map222 commented on python/pyspark/sql/column.py:
```diff
@@ -250,11 +250,39 @@ def __iter__(self):
         raise TypeError("Column is not iterable")
 
     # string methods
+    _rlike_doc = """ Return a Boolean :class:`Column` based on a regex match.\n
```
Without the newlines, the description and and params lines are combined into one line. I don't totally understand why.

HyukjinKwon replied:
I just manually tried after replacing `\n` to a newline with `make clean` and `make html`. It seems fine to use a newline (not `\n`). Could you double check and replace it if I was not wrong?

map222 replied:
I made a branch without the newlines in `rlike` and `like`. This is the html it makes: https://raw.githubusercontent.com/map222/spark/50fc4f4a4a19a95eb5eaae76840d63f540bd45e0/python/docs/_build/html/pyspark.sql.html
In the html, the `param` line is concatenated to the first one.
Could there be a setting on my local machine that is different?

HyukjinKwon replied:
Could you maybe give a shot with this patch - https://github.com/map222/spark/compare/patterson-documentation...HyukjinKwon:rlike-docstring.patch

?

I double checked it produces 

![2017-04-04 1 23 30](https://cloud.githubusercontent.com/assets/6477701/24641412/84765e9c-193a-11e7-85d5-9745ea151c12.png)

map222 replied:
That fixed it!

map222 committed:
Added docs for isNull and isNotNull

HyukjinKwon commented on python/pyspark/sql/column.py:
```diff
@@ -250,11 +250,41 @@ def __iter__(self):
+    _rlike_doc = """ Return a Boolean :class:`Column` based on a regex match.\n
+                     :param other: an extended regex expression
+
+                     >>> df.filter( df.name.rlike('ice$') ).collect()
```
Let's make sure there is no extra space between `(...)`.

HyukjinKwon reviewed (commented):
It looks quite good to me except for few comments. These are all from me.

HyukjinKwon commented on python/pyspark/sql/column.py:
```diff
@@ -303,8 +333,25 @@ def isin(self, *cols):
+                      >>> df2.filter( df2.height.isNull ).collect()
+                      [Row(name=u'Alice', height=None)]
+                  '''
+    _isNotNull_doc = ''' True if the current expression is null. Often combined with 
```
BTW, just a question. Do we need the leading space here in the documentation? I think we should remove if it is unnecessary.

map222 replied:
Do you mean formatting it more like https://github.com/map222/spark/blob/90d1150867184d86b029c1f6397dcd855b1f5961/python/pyspark/sql/column.py#L482-L493?

HyukjinKwon replied:
Up to my knowledge, both docstrings comply pep8 up to my knowledge,
```
""" ...
"""
```

or 

```
"""
...
"""
```

but for this case, it seems a separate variable. Personally, I prefer

```python
    _isNull_doc = """
    True if the current expression is null. Often combined with
    :func:`DataFrame.filter` to select rows with null values.

    >>> df2.collect()
    [Row(name=u'Tom', height=80), Row(name=u'Alice', height=None)]
    >>> df2.filter( df2.height.isNull ).collect()
    [Row(name=u'Alice', height=None)]
    """
```

but I could not find a formal reference to support this idea (in case that it is a separate variable) and I am not supposed to decide this. So, I am fine.

HyukjinKwon replied:
^ cc @holdenk

HyukjinKwon commented:
Also, you could run the python tests via `./python/run-tests.py --module pyspark-sql` for python tests after building. It is fine because I guess Jenkins will catch this but it might be nicer if they could be tested in the local ahead if possible.

map222 committed:
Tuned whitespace

HyukjinKwon commented on python/pyspark/sql/column.py:
```diff
@@ -303,8 +333,25 @@ def isin(self, *cols):
 
-    isNull = _unary_op("isNull", "True if the current expression is null.")
-    isNotNull = _unary_op("isNotNull", "True if the current expression is not null.")
+    _isNull_doc = ''' True if the current expression is null. Often combined with 
```
I just found a good reference in pep8

> For triple-quoted strings, always use double quote characters to be consistent with the docstring convention in PEP 257

https://www.python.org/dev/peps/pep-0008/#string-quotes

HyukjinKwon commented:
It might be better to run`./dev/lint-python` locally if possible. There will catch more of minor nits ahead.

map222 committed:
Fixed formatting issues: parentheses, block quotes, and newline

map222 commented:
I think the latest commit addresses the formatting issues from above: removed spaces insides `(..)`, removed the `\n` newlines, and made the blockquotes more consistent with the rest of the code.

I also ran `./dev/lint-python`, and fixed some trailing whitespace issues. All pydoc checks passed.

holdenk commented:
Jenkins OK to test.

map222 commented:
@HyukjinKwon Do I need to do something to start the Jenkins test?

HyukjinKwon commented:
It seems somethine goes wrong with @holdnk and Jenkins. I think I dont have a permission to trigger this.

holdenk commented:
I don't know, let me ping @joshrosen and see what's going on.

Jenkins retest this please.

holdenk commented:
I'll take a look at Jenkins and see if I can figure out why it isn't working - but I don't have a lot of experience coaxing Jenkins.

HyukjinKwon commented:
It is little bit funny that some contibutors appearntly are able to trigger the build (at least I saw a contributor made a comment "retest this please" and it triggers the build) whereas some committers are not. Did I misunderstand? I saw a comment that @felixcheung is actually unable to trigger the build too if I am not mistaken.

holdenk commented:
No your correct. The tooling around Jenkins hasn't had enough love as of late since there are plans to replace a lot of it, so newer committers aren't always added everywhere they need to be. I've got some access I can look and see if I can fix it, but if I don't see where we'll have to wait for Josh or Shane (who have been very helpful) to update the config.

holdenk commented:
Jenkins retest this please

holdenk commented:
I've e-mailed them since the Jenkins configuration is a bit too involved (and I'd need Shane to sign off on any Jenkins change anyways). Sorry this is slowing down your PR @map222 and thank you so much for your patience with us :)

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #75622 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75622/testReport)** for PR 17469 at commit [`bc03f3c`](https://github.com/apache/spark/commit/bc03f3c5799e749558696fef0723e592523fbcd9).

HyukjinKwon commented:
retest this please

HyukjinKwon commented:
Yes, it seems from your comment @holdenk.

holdenk commented:
Great! I'll still follow up with Shane & Josh re: @felixcheung triggering build as well.

SparkQA commented:
**[Test build #75622 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75622/testReport)** for PR 17469 at commit [`bc03f3c`](https://github.com/apache/spark/commit/bc03f3c5799e749558696fef0723e592523fbcd9).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75622/
Test FAILed.

map222 commented:
@HyukjinKwon The Jenkins test failed. I'm having trouble running the tests locally (I can't build Spark yet), and I can't decipher the Jenkins error messages. Does something jump out to you?

HyukjinKwon commented on python/pyspark/sql/column.py:
```diff
@@ -250,11 +250,50 @@ def __iter__(self):
+    :param other: an extended regex expression
+
+    >>> df.filter(df.name.rlike('ice$')).collect()
+    [Row(name=u'Alice', age=1)]
```
It sounds the test failure is here. If you click the link, the full logs can be checked (e.g., https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75622/console).

Up to my knowledge, Row in python sorts the field names and therefore I guess it should be `[Row(age=1, name=u'Alice')]`.

HyukjinKwon commented on python/pyspark/sql/column.py:
```diff
@@ -303,8 +342,27 @@ def isin(self, *cols):
+    True if the current expression is null. Often combined with
+    :func:`DataFrame.filter` to select rows with null values.
+
+    >>> df2.collect()
```
Also, It seems somehow it can't fine `df2` here in the doctest just as the error message says.

map222 committed:
Fixed documentation to match unit tests

map222 commented:
@HyukjinKwon I have updated the documentation to pass the tests (I didn't realize the tests actually executed the documented code!). I was able to build Spark locally, and run `./python/run-tests.py --module pyspark-sql` which passed.

HyukjinKwon commented:
@map222, thanks, I don't have a permission to retrigger the build and know why it does not. Would be great if @srowen or @holdenk are able to retrigger the build.

HyukjinKwon commented:
@map222, thanks, I don't have a permission to retrigger the build and know why it does not automatically from the newly pushed commit. Would be great if @srowen or @holdenk are able to retrigger the build.

JoshRosen commented:
jenkins retest this please

SparkQA commented:
**[Test build #75723 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75723/testReport)** for PR 17469 at commit [`b157bc3`](https://github.com/apache/spark/commit/b157bc3288c326fac847aca3fecb8d0f79592f42).

SparkQA commented:
**[Test build #75723 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75723/testReport)** for PR 17469 at commit [`b157bc3`](https://github.com/apache/spark/commit/b157bc3288c326fac847aca3fecb8d0f79592f42).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75723/
Test FAILed.

holdenk commented:
I've added @felixcheung to Jenkins as a GHPRB admin so now he should be able to trigger Jenkins requests as well :)

map222 committed:
Added imports and fixed functions

map222 commented:
I can't get the test failure to replicate on my local machine. Running `./python/run-tests.py --modules=pyspark-sql` locally doesn't give any errors. The previous failure was due to not having `import Row`, which I added.

HyukjinKwon commented:
It looks good to me if it passes the test. I could request filling up other documentation as well but if it is difficult to run the tests in local machine and regarding (I assume) it is the very first contribution, I think it is probably fine.

Maybe, I can make a followup to fill up all if all are fine here with it. I checked the documentation output manually before.

HyukjinKwon commented:
I think we need @holdenk's sign-off though.

map222 commented:
I'd be happy to add other documentation, but I would also like to get my first commit in! If you make a separate ticket to track that would help.

HyukjinKwon commented:
Usually, it more makes sense to handle the similar instances in one place. If you are easily able to add them, probably, it is a good place to put them all. If it is currently difficult, I think it is fine as is (for the reason above) . 

If other committers here or @holdenk, think these should be handled all together here (usually they are), let's add all here.

map222 commented:
@holdenk @srowen Could I get a Jenkins test for this?

felixcheung commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #75955 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75955/testReport)** for PR 17469 at commit [`e785dbc`](https://github.com/apache/spark/commit/e785dbc9bd32a45d0e415f58ee8446cd03276b10).

SparkQA commented:
**[Test build #75955 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75955/testReport)** for PR 17469 at commit [`e785dbc`](https://github.com/apache/spark/commit/e785dbc9bd32a45d0e415f58ee8446cd03276b10).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75955/
Test FAILed.

map222 commented:
The documentation unit test is now failing due to differences in python 2 and 3 strings. This is one of the error messages using the python 3.4 kernel:
```File "/home/jenkins/workspace/SparkPullRequestBuilder/python/pyspark/sql/column.py", line 122, in pyspark.sql.column.Column.endswith
Failed example:
    df.filter(df.name.endswith('ice')).collect()
Differences (ndiff with -expected +actual):
    - [Row(age=2, name=u'Alice')]
    ?                  -
    + [Row(age=2, name='Alice')]```

I don't know how to fix this. I don't want to redefine `df` for each example. It looks like there is a function in the RDD to convert string types, https://github.com/apache/spark/blob/master/python/pyspark/rdd.py#L146

zero323 commented:
@map222 [`ignore_unicode_prefix`](https://github.com/apache/spark/blob/8ddf0d2a60795a2306f94df8eac6e265b1fe5230/python/pyspark/rdd.py#L146-L156)

map222 commented:
@zero323 Aaaah, I had even identified what I needed to do! So I just need to decorate `_unary_op` and `_bin_op`, yes?

map222 committed:
Added ignore_unicode_prefix to avoid string error

zero323 commented:
You'll have to use it on the wrapped function. Like:

    contains = ignore_unicode_prefix(_bin_op("contains"))

map222 commented:
Ok, I added ignore_unicode_prefix to the 6 functions, and it passed local tests. I think it is ready for Jenkins again.

felixcheung commented:
Jenkins, retest this please

felixcheung commented:
I don't why Jenkins doesn't pick up the changes automatically...

SparkQA commented:
**[Test build #76075 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76075/testReport)** for PR 17469 at commit [`b52765f`](https://github.com/apache/spark/commit/b52765f5ef156862bd3cc4793a0d3fbd4d334449).

SparkQA commented:
**[Test build #76075 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76075/testReport)** for PR 17469 at commit [`b52765f`](https://github.com/apache/spark/commit/b52765f5ef156862bd3cc4793a0d3fbd4d334449).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76075/
Test PASSed.

HyukjinKwon commented:
LGTM if committers are okay with merging fixing some of documentation (not all) but regarding it is his very first contribution.

felixcheung commented:
@holdenk

holdenk commented:
LGTM, thanks for your work on this @map222 & thanks for your work reviewing this @HyukjinKwon. 
Merged to master.

liancheng commented:
Also cherry-picked this to branch-2.2.

cc @gatorsmile
